docs/design_proposals/configurable-kubecontext.md

```diff
@@ -0,0 +1,183 @@
+
+The goal of this document is to create an agreement on what options should be supported and identify edge cases.
+
+## Design
+
+There are four places where kubecontext activation can be added:
+<table>
+    <thead>
+        <th>Precedence</th> <th>Kind</th> <th>Details</th>
+    </thead>
+    <tbody>
+        <tr>
+            <td>1. (highest)</td>
+            <td>CLI option</td>
+            <td>
+              The kubernetes standard to set the kubecontext is <code>--context</code>.
+              However, in Skaffold this term is so overloaded that it should more precisely be named <code>--kube-context</code>.
+              This flag is necessary for IDE integration.
+            </td>
+        </tr>
+        <tr>
+            <td>2.</td>
+            <td>Env variable</td>
+            <td>
+              <code>SKAFFOLD_KUBE_CONTEXT</code>, similar to other Skaffold flags.
+            </td>
+        </tr>
+        <tr>
+            <td>3.</td>
+            <td><code>skaffold.yaml</code></td>
+            <td>
+              Json-path <code>deploy.kubeContext</code>.
+              This option is shareable, but also creates a lot of complexity due to profile activation (see detailed discussion below).
+            </td>
+        </tr>
+        <tr>
+            <td>4. (lowest)</td>
+            <td>Global Skaffold config</td>
+            <td>
+              This should give users the possibility to define a default context globally or per project.
+              This variant is not shareable.
+            </td>
+        </tr>
+    </tbody>
+</table>
+
+---
+
+Beside the kubecontext, also the namespace needs to be specified.
+It should also be possible to specify in various places:
+
+1. `--namespace` CLI flag
+2. `SKAFFOLD_NAMESPACE` env var
+3. In `skaffold.yaml` at json-path `deploy.namespace`
+4. globally or per project in the Skaffold config
+
+### Detailed discussion
+#### Option in `skaffold.yaml`
+A configuration option in `skaffold.yaml` has the advantage of being most discoverable:
+it is in the place where users configure all aspects of their pipeline.
+In addition, it allows to define the kubecontext per Skaffold profile.
+
+However, it also has questionable implications:
+
+- `skaffold.yaml` is meant to be shared, but kubecontext names vary across users.
+  Sharing therefore makes only sense in a corporate setting where context names are the same across many users.
+  There is however a risk of abuse in settings where sharing the context name does not make sense, for example in open source projects.
+- Due to profile activation by Skaffold profiles, there can be circular problems.
+  For example, the current context is `A` and activates some profile.
+  This profile activates a different kubecontext `B`.
+  A solution could be to forbid specifying a kubecontext in a profile if it is activated by a kubecontext (and validate that).
```

Comment:
This should be the recommended solution. 
We should mention that this check should on the "effective profile"  that is the result of merging all the activated profiles.

Reply:
I'm a bit puzzled about the effective profile. If we forbid to override the kubecontext twice by two profiles, then we are deviating from the normal behavior of how profiles are usually applied. Usually profiles are applied one after the other, and last one wins. So why should that be different for the kubecontext?

I think a static check where a profile with activation by kubecontext must not override the kubecontext is easy to understand and fine. However, as soon as the logic becomes dynamic this quickly becomes rather intransparent.

Reply:
Hm...but then 

given: current context = minikube 
activated profiles: 

P1 <- envvar1
P2 <- cur_ctx=minikube  

and P1.deploy.kubecontext = gke_foobar

I'll end up with the minikube specific profile that deploys to gke...

Reply:
I would say

given: current context = minikube

P1 <- envvar1,   changes the kubeContext to `gke_foobar`
P2 _no longer applies_, because kubeContext is not `minikube` anymore

---

I just checked the `skaffold.yaml` and the profile activation by kube-context is simply called `kubeContext`. It does not necessarily mean that it only considers the current context as defined in the kubeconfig. That would even be very misleading: when somebody changes the kube-context on the CLI with `skaffold --kube-context gke_foobar`, but the kubeconfig says `minikube`, I would certainly not expect that Skaffold activates a minikube profile.

Reply:
I think profile activation is confusing, and one can construct pretty contrived examples:

given: current context = `minikube`
P1 <- envvar, changes kubeContext to `gke_foobar`
P2 <- activated by kubeContext `gke_foobar`
P3 <- activated by envvar, changes kubeContext to `minikube`

In that sequence, the gke-specific profile would be activated on minikube. :zap: 

---

This example suggests that a dynamic check is necessary in any case. I think the validation should look out for profiles that were activated by a kube-context which is no longer active in the effective `skaffold.yaml` (Pardon me, if that is what you were trying to tell me all along).

Reply:
Never mind my previous comments. I should have read the code first.

So profile activation and profile application are separate, meaning that all profile activations use the same current kubecontext. Iow there can be no transitive kubecontext activation like
```
P1 activates minikube
P2 is activated by minikube
...and so on
```
A reasonable validation is therefore:
- check if any profile is activated by the current kubecontext before profile activation
- after profile application, check if the effective kubecontext `skaffold.yaml` has changed.
- if yes, and there was a context-activated profile, return error
- otherwise, everything is fine.
(Pardon me, if this is what you wanted to tell me all along..)
This check also makes any static check of the `skaffold.yaml` obsolete.

The case where subsequent profiles repeatedly switch the kubecontext is probably fine, as profile activation follows a last-one-wins policy which should also apply to kubecontext.

Reply:
This sounds good, yes, this is exactly what I meant :)